Debezium errors with postgresql CDC

### Description

The following errors are printed for the CDC-enabled PostgreSQL service. 


```
Running executable

error: Function 'onRead' is not available. {"payload":{"before":null,"after":{"id":1,"name":"John Doe","email":"john.doe@example.com","created_at":1776835977040843},"source":{"version":"3.0.8.Final","connector":"postgresql","name":"bal_cdc_schema_history","ts_ms":1776837124512,"snapshot":"first","db":"testdb","sequence":"[null,"26713952"]","ts_us":1776837124512182,"ts_ns":1776837124512182000,"schema":"public","table":"customers","txId":749,"lsn":26713952,"xmin":null},"transaction":null,"op":"r","ts_ms":1776837125140,"ts_us":1776837125140728,"ts_ns":1776837125140728000}}
error: Function 'onRead' is not available. {"payload":{"before":null,"after":{"id":2,"name":"Jane Smith","email":"jane.smith@example.com","created_at":1776835977040843},"source":{"version":"3.0.8.Final","connector":"postgresql","name":"bal_cdc_schema_history","ts_ms":1776837124512,"snapshot":"true","db":"testdb","sequence":"[null,"26713952"]","ts_us":1776837124512182,"ts_ns":1776837124512182000,"schema":"public","table":"customers","txId":749,"lsn":26713952,"xmin":null},"transaction":null,"op":"r","ts_ms":1776837125141,"ts_us":1776837125141865,"ts_ns":1776837125141865000}}
error: Function 'onRead' is not available. {"payload":{"before":null,"after":{"id":3,"name":"Bob Johnson","email":"bob.johnson@example.com","created_at":1776835977040843},"source":{"version":"3.0.8.Final","connector":"postgresql","name":"bal_cdc_schema_history","ts_ms":1776837124512,"snapshot":"true","db":"testdb","sequence":"[null,"26713952"]","ts_us":1776837124512182,"ts_ns":1776837124512182000,"schema":"public","table":"customers","txId":749,"lsn":26713952,"xmin":null},"transaction":null,"op":"r","ts_ms":1776837125142,"ts_us":1776837125142005,"ts_ns":1776837125142005000}}
error: Function 'onRead' is not available. {"payload":{"before":null,"after":{"id":4,"name":"Alice Williams","email":"alice.williams@example.com","created_at":1776836039595845},"source":{"version":"3.0.8.Final","connector":"postgresql","name":"bal_cdc_schema_history","ts_ms":1776837124512,"snapshot":"true","db":"testdb","sequence":"[null,"26713952"]","ts_us":1776837124512182,"ts_ns":1776837124512182000,"schema":"public","table":"customers","txId":749,"lsn":26713952,"xmin":null},"transaction":null,"op":"r","ts_ms":1776837125142,"ts_us":1776837125142151,"ts_ns":1776837125142151000}}
error: Function 'onRead' is not available. {"payload":{"before":null,"after":{"id":5,"name":"Michael Brown","email":"michael.brown@example.com","created_at":1776836420308043},"source":{"version":"3.0.8.Final","connector":"postgresql","name":"bal_cdc_schema_history","ts_ms":1776837124512,"snapshot":"true","db":"testdb","sequence":"[null,"26713952"]","ts_us":1776837124512182,"ts_ns":1776837124512182000,"schema":"public","table":"customers","txId":749,"lsn":26713952,"xmin":null},"transaction":null,"op":"r","ts_ms":1776837125142,"ts_us":1776837125142271,"ts_ns":1776837125142271000}}
error: Function 'onRead' is not available. {"payload":{"before":null,"after":{"id":6,"name":"Sarah Davis","email":"sarah.davis@example.com","created_at":1776836567109229},"source":{"version":"3.0.8.Final","connector":"postgresql","name":"bal_cdc_schema_history","ts_ms":1776837124512,"snapshot":"last","db":"testdb","sequence":"[null,"26713952"]","ts_us":1776837124512182,"ts_ns":1776837124512182000,"schema":"public","table":"customers","txId":749,"lsn":26713952,"xmin":null},"transaction":null,"op":"r","ts_ms":1776837125142,"ts_us":1776837125142392,"ts_ns":1776837125142392000}}

time=2026-04-22T11:23:16.727+05:30 level=INFO module=anuruddha/cdcpostgresql message="customers{\"id\":7,\"name\":\"David Wilson\",\"email\":\"david.wilson@example.com\",\"created_at\":1776837195098306}"
Apr 22, 2026 11:23:16 AM io.debezium.embedded.EmbeddedEngine commitOffsets
SEVERE: Flush of EmbeddedEngine{id=ballerina-cdc-connector} offsets threw an unexpected exception: 
java.util.concurrent.ExecutionException: org.apache.kafka.connect.errors.ConnectException: java.nio.file.NoSuchFileException: tmp/debezium-offsets.dat
        at java.base/java.util.concurrent.FutureTask.report(FutureTask.java:122)
        at java.base/java.util.concurrent.FutureTask.get(FutureTask.java:205)
        at io.debezium.embedded.EmbeddedEngine.commitOffsets(EmbeddedEngine.java:915)
        at io.debezium.embedded.EmbeddedEngine.maybeFlush(EmbeddedEngine.java:891)
        at io.debezium.embedded.EmbeddedEngine$4.markBatchFinished(EmbeddedEngine.java:830)
        at io.debezium.embedded.ConvertingEngineBuilder$ConvertingChangeConsumer$1.markBatchFinished(ConvertingEngineBuilder.java:87)
        at io.ballerina.lib.cdc.BalChangeConsumer.handleBatch(BalChangeConsumer.java:134)
        at io.debezium.embedded.ConvertingEngineBuilder$ConvertingChangeConsumer.handleBatch(ConvertingEngineBuilder.java:75)
        at io.debezium.embedded.EmbeddedEngine.pollRecords(EmbeddedEngine.java:731)
        at io.debezium.embedded.EmbeddedEngine.run(EmbeddedEngine.java:470)
        at io.debezium.embedded.ConvertingEngineBuilder$1.run(ConvertingEngineBuilder.java:163)
        at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:572)
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)
        at java.base/java.lang.VirtualThread.run(VirtualThread.java:329)
Caused by: org.apache.kafka.connect.errors.ConnectException: java.nio.file.NoSuchFileException: tmp/debezium-offsets.dat
        at org.apache.kafka.connect.storage.FileOffsetBackingStore.save(FileOffsetBackingStore.java:109)
        at org.apache.kafka.connect.storage.MemoryOffsetBackingStore.lambda$set$1(MemoryOffsetBackingStore.java:81)
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
        at java.base/java.lang.Thread.run(Thread.java:1583)
Caused by: java.nio.file.NoSuchFileException: tmp/debezium-offsets.dat
        at java.base/sun.nio.fs.UnixException.translateToIOException(UnixException.java:92)
        at java.base/sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:106)
        at java.base/sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:111)
        at java.base/sun.nio.fs.UnixFileSystemProvider.newByteChannel(UnixFileSystemProvider.java:261)
        at java.base/java.nio.file.spi.FileSystemProvider.newOutputStream(FileSystemProvider.java:482)
        at java.base/java.nio.file.Files.newOutputStream(Files.java:228)
        at org.apache.kafka.connect.storage.FileOffsetBackingStore.save(FileOffsetBackingStore.java:99)
        ... 5 more

time=2026-04-22T11:23:43.457+05:30 level=INFO module=anuruddha/cdcpostgresql message="Capturing update from table customers: {\"id\":1,\"name\":\"John Doe\",\"email\":\"john.doe.updated@example.com\",\"created_at\":1776835977040843}"
Apr 22, 2026 11:23:43 AM io.debezium.embedded.EmbeddedEngine commitOffsets
SEVERE: Flush of EmbeddedEngine{id=ballerina-cdc-connector} offsets threw an unexpected exception: 
java.util.concurrent.ExecutionException: org.apache.kafka.connect.errors.ConnectException: java.nio.file.NoSuchFileException: tmp/debezium-offsets.dat
        at java.base/java.util.concurrent.FutureTask.report(FutureTask.java:122)
        at java.base/java.util.concurrent.FutureTask.get(FutureTask.java:205)
        at io.debezium.embedded.EmbeddedEngine.commitOffsets(EmbeddedEngine.java:915)
        at io.debezium.embedded.EmbeddedEngine.maybeFlush(EmbeddedEngine.java:891)
        at io.debezium.embedded.EmbeddedEngine$4.markBatchFinished(EmbeddedEngine.java:830)
        at io.debezium.embedded.ConvertingEngineBuilder$ConvertingChangeConsumer$1.markBatchFinished(ConvertingEngineBuilder.java:87)
        at io.ballerina.lib.cdc.BalChangeConsumer.handleBatch(BalChangeConsumer.java:134)
        at io.debezium.embedded.ConvertingEngineBuilder$ConvertingChangeConsumer.handleBatch(ConvertingEngineBuilder.java:75)
        at io.debezium.embedded.EmbeddedEngine.pollRecords(EmbeddedEngine.java:731)
        at io.debezium.embedded.EmbeddedEngine.run(EmbeddedEngine.java:470)
        at io.debezium.embedded.ConvertingEngineBuilder$1.run(ConvertingEngineBuilder.java:163)
        at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:572)
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)
        at java.base/java.lang.VirtualThread.run(VirtualThread.java:329)
Caused by: org.apache.kafka.connect.errors.ConnectException: java.nio.file.NoSuchFileException: tmp/debezium-offsets.dat
        at org.apache.kafka.connect.storage.FileOffsetBackingStore.save(FileOffsetBackingStore.java:109)
        at org.apache.kafka.connect.storage.MemoryOffsetBackingStore.lambda$set$1(MemoryOffsetBackingStore.java:81)
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
        at java.base/java.lang.Thread.run(Thread.java:1583)
Caused by: java.nio.file.NoSuchFileException: tmp/debezium-offsets.dat
        at java.base/sun.nio.fs.UnixException.translateToIOException(UnixException.java:92)
        at java.base/sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:106)
        at java.base/sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:111)
        at java.base/sun.nio.fs.UnixFileSystemProvider.newByteChannel(UnixFileSystemProvider.java:261)
        at java.base/java.nio.file.spi.FileSystemProvider.newOutputStream(FileSystemProvider.java:482)
        at java.base/java.nio.file.Files.newOutputStream(Files.java:228)
        at org.apache.kafka.connect.storage.FileOffsetBackingStore.save(FileOffsetBackingStore.java:99)
        ... 5 more
```

### Steps to Reproduce

Create a CDC-enabled PostgreSQL database and execute the following code from WSO2 Integrator. 

```ballerina
import ballerina/log;
import ballerinax/cdc;
import ballerinax/postgresql;
import ballerinax/postgresql.cdc.driver as _;

listener postgresql:CdcListener postgresqlCdcListener = new (database = {hostname: "localhost", port: 5432, username: "postgres", password: "postgres", databaseName: "testdb"});

@cdc:ServiceConfig {
    tables: "testdb.public.customers"
}

service cdc:Service on postgresqlCdcListener {
    remote function onCreate(record {} afterEntry, string tableName) returns error? {
        do {
            log:printInfo(string `${tableName}${afterEntry.toString()}`);
        } on fail error err {
            // handle error
            return error("unhandled error", err);
        }
    }

    remote function onUpdate(record {} beforeEntry, record {} afterEntry, string tableName) returns error? {
        do {
            log:printInfo(string `Capturing update from table ${tableName}: ${afterEntry.toString()}`);
        } on fail error err {
            // handle error
            return error("unhandled error", err);
        }
    }
}

```

Used the below docker setup to create postgresql.

[docker-compose-postgres.yml](https://github.com/user-attachments/files/26956291/docker-compose-postgres.yml)
[init-postgres.sql](https://github.com/user-attachments/files/26956292/init-postgres.sql)

### Version

Version: 5.0.0-alpha7 OS: Darwin arm64 25.4.0

### Environment Details (with versions)

MacOS

--- a/ballerina/Ballerina.toml
+++ b/ballerina/Ballerina.toml
@@ -4,6 +4,7 @@ name = "cdc"
 version = "1.3.0"
 distribution = "2201.12.0"
 authors = ["Ballerina"]
+keywords = ["Type/Connector", "Type/Library", "Area/Database", "Vendor/Red Hat", "Debezium"]
 repository = "https://github.com/ballerina-platform/module-ballerinax-cdc"
 license = ["Apache-2.0"]
 

--- a/ballerina/Dependencies.toml
+++ b/ballerina/Dependencies.toml
@@ -5,12 +5,12 @@
 
 [ballerina]
 dependencies-toml-version = "2"
-distribution-version = "2201.13.1"
+distribution-version = "2201.12.0"
 
 [[package]]
 org = "ballerina"
 name = "crypto"
-version = "2.11.0"
+version = "2.9.3"
 dependencies = [
 	{org = "ballerina", name = "jballerina.java"},
 	{org = "ballerina", name = "time"}
@@ -34,7 +34,7 @@ modules = [
 [[package]]
 org = "ballerina"
 name = "file"
-version = "1.13.0"
+version = "1.12.0"
 scope = "testOnly"
 dependencies = [
 	{org = "ballerina", name = "io"},
@@ -142,7 +142,7 @@ dependencies = [
 [[package]]
 org = "ballerina"
 name = "log"
-version = "2.17.0"
+version = "2.12.0"
 dependencies = [
 	{org = "ballerina", name = "io"},
 	{org = "ballerina", name = "jballerina.java"},
@@ -156,7 +156,7 @@ modules = [
 [[package]]
 org = "ballerina"
 name = "observe"
-version = "1.7.1"
+version = "1.5.1"
 dependencies = [
 	{org = "ballerina", name = "jballerina.java"}
 ]
@@ -187,7 +187,7 @@ modules = [
 [[package]]
 org = "ballerina"
 name = "sql"
-version = "1.19.0"
+version = "1.16.0"
 scope = "testOnly"
 dependencies = [
 	{org = "ballerina", name = "io"},
@@ -255,7 +255,7 @@ modules = [
 [[package]]
 org = "ballerinax"
 name = "java.jdbc"
-version = "1.15.1"
+version = "1.14.1"
 scope = "testOnly"
 dependencies = [
 	{org = "ballerina", name = "io"},

--- a/ballerina/Dependencies.toml
+++ b/ballerina/Dependencies.toml
@@ -5,12 +5,12 @@
 
 [ballerina]
 dependencies-toml-version = "2"
-distribution-version = "2201.12.0"
+distribution-version = "2201.13.1"
 
 [[package]]
 org = "ballerina"
 name = "crypto"
-version = "2.9.3"
+version = "2.11.0"
 dependencies = [
 	{org = "ballerina", name = "jballerina.java"},
 	{org = "ballerina", name = "time"}
@@ -34,7 +34,7 @@ modules = [
 [[package]]
 org = "ballerina"
 name = "file"
-version = "1.12.0"
+version = "1.13.0"
 scope = "testOnly"
 dependencies = [
 	{org = "ballerina", name = "io"},
@@ -142,7 +142,7 @@ dependencies = [
 [[package]]
 org = "ballerina"
 name = "log"
-version = "2.12.0"
+version = "2.17.0"
 dependencies = [
 	{org = "ballerina", name = "io"},
 	{org = "ballerina", name = "jballerina.java"},
@@ -156,7 +156,7 @@ modules = [
 [[package]]
 org = "ballerina"
 name = "observe"
-version = "1.5.1"
+version = "1.7.1"
 dependencies = [
 	{org = "ballerina", name = "jballerina.java"}
 ]
@@ -187,7 +187,7 @@ modules = [
 [[package]]
 org = "ballerina"
 name = "sql"
-version = "1.16.0"
+version = "1.19.0"
 scope = "testOnly"
 dependencies = [
 	{org = "ballerina", name = "io"},
@@ -255,7 +255,7 @@ modules = [
 [[package]]
 org = "ballerinax"
 name = "java.jdbc"
-version = "1.14.1"
+version = "1.15.1"
 scope = "testOnly"
 dependencies = [
 	{org = "ballerina", name = "io"},
@@ -271,7 +271,7 @@ modules = [
 [[package]]
 org = "ballerinax"
 name = "mysql.cdc.driver"
-version = "1.0.1"
+version = "1.0.2"
 scope = "testOnly"
 dependencies = [
 	{org = "ballerinax", name = "mysql.driver"}
@@ -283,6 +283,6 @@ modules = [
 [[package]]
 org = "ballerinax"
 name = "mysql.driver"
-version = "1.8.0"
+version = "1.8.1"
 scope = "testOnly"
 

--- a/ballerina/tests/dynamic_attachment.bal
+++ b/ballerina/tests/dynamic_attachment.bal
@@ -14,6 +14,7 @@
 // specific language governing permissions and limitations
 // under the License.
 
+import ballerina/file;
 import ballerina/test;
 import ballerina/lang.runtime;
 
@@ -215,4 +216,18 @@ function testLivenessWithoutReceivingEvents() returns error? {
     boolean liveness = check isLive(mysqlListener);
     test:assertFalse(liveness, "Liveness check passes even after not receiving events within the liveness interval");
     check mysqlListener.gracefulStop();
+}
+
+@test:Config {}
+function testOffsetStorageParentDirCreation() returns error? {
+    MockListener mysqlListener = new ({
+        database: {username, password, port},
+        options: {snapshotMode: NO_DATA},
+        offsetStorage: {fileName: "tmp/nested/offsets.dat"}
+    });
+    check mysqlListener.attach(testService);
+    check mysqlListener.'start();
+    test:assertTrue(check file:test("tmp/nested", file:EXISTS),
+            msg = "Parent directory for offset storage file was not created.");
+    check mysqlListener.immediateStop();
 }

--- a/ballerina/tests/listener_tests.bal
+++ b/ballerina/tests/listener_tests.bal
@@ -112,7 +112,7 @@ function testMockListenerEvents() returns error? {
         `INSERT INTO vendors (id, name, contact_info) 
         VALUES (201, 'Vendor A', 'contact@vendora.com')`);
     runtime:sleep(3);
-    test:assertEquals(onErrorCount, 3, msg = "Error count mismatch.");
+    test:assertEquals(onErrorCount, 1, msg = "Error count mismatch.");
     // 1,2 for onRead method not present, 3 for payload binding failure
 
     check testListener.gracefulStop();

--- a/ballerina/tests/listener_tests.bal
+++ b/ballerina/tests/listener_tests.bal
@@ -112,8 +112,8 @@ function testMockListenerEvents() returns error? {
         `INSERT INTO vendors (id, name, contact_info) 
         VALUES (201, 'Vendor A', 'contact@vendora.com')`);
     runtime:sleep(3);
+    // Missing onRead on vendors triggers a warning, not an error — only the binding failure reaches onError
     test:assertEquals(onErrorCount, 1, msg = "Error count mismatch.");
-    // 1,2 for onRead method not present, 3 for payload binding failure
 
     check testListener.gracefulStop();
 }

--- a/native/src/main/java/io/ballerina/lib/cdc/BalChangeConsumer.java
+++ b/native/src/main/java/io/ballerina/lib/cdc/BalChangeConsumer.java
@@ -41,6 +41,7 @@ import io.ballerina.runtime.api.values.BTypedesc;
 import io.debezium.engine.ChangeEvent;
 import io.debezium.engine.DebeziumEngine;
 
+import java.io.PrintStream;
 import java.time.Instant;
 import java.util.ArrayList;
 import java.util.HashMap;
@@ -67,6 +68,7 @@ import static java.lang.Boolean.FALSE;
  * Handles change events from the Debezium engine and invokes the appropriate Ballerina service methods.
  */
 public class BalChangeConsumer implements DebeziumEngine.ChangeConsumer<ChangeEvent<String, String>> {
+    private static final PrintStream ERR_OUT = System.err;
 
     private final Map<String, Service> serviceMap;
     private final boolean isSingleServiceAttached;
@@ -214,7 +216,7 @@ public class BalChangeConsumer implements DebeziumEngine.ChangeConsumer<ChangeEv
                         ServiceMethodNames.ON_ERROR, metaData, bError);
                 handleReturnValue(returnValue);
             } else {
-                bError.printStackTrace();
+                ERR_OUT.println("warning: " + bError.getPrintableStackTrace());
             }
         } catch (BError balError) {
             balError.printStackTrace();

--- a/native/src/main/java/io/ballerina/lib/cdc/BalChangeConsumer.java
+++ b/native/src/main/java/io/ballerina/lib/cdc/BalChangeConsumer.java
@@ -109,7 +109,9 @@ public class BalChangeConsumer implements DebeziumEngine.ChangeConsumer<ChangeEv
                 String methodName = getMethodName(payload.getOp());
                 Method method = selectedService.getMethod(methodName);
                 if (method == null) {
-                    throw createMethodNotFoundError(payload, methodName);
+                    ERR_OUT.println("warning: Function '" + methodName + "' is not available.");
+                    committer.markProcessed(record);
+                    continue;
                 }
 
                 boolean isIsolated = selectedService.isIsolated() && method.isIsolated();
@@ -169,10 +171,6 @@ public class BalChangeConsumer implements DebeziumEngine.ChangeConsumer<ChangeEv
 
     private Object[] processParameters(Service service, String functionName, Payload payload) {
         Method method = service.getMethod(functionName);
-        if (method == null) {
-            throw createMethodNotFoundError(payload, functionName);
-        }
-
         List<Object> parameters = new ArrayList<>();
         if (method.hasBeforeParam()) {
             parameters.add(processParameterToIntendedType(payload, EventMembers.BEFORE, method.beforeParamType()));
@@ -216,7 +214,7 @@ public class BalChangeConsumer implements DebeziumEngine.ChangeConsumer<ChangeEv
                         ServiceMethodNames.ON_ERROR, metaData, bError);
                 handleReturnValue(returnValue);
             } else {
-                ERR_OUT.println("warning: " + bError.getPrintableStackTrace());
+                bError.printStackTrace();
             }
         } catch (BError balError) {
             balError.printStackTrace();
@@ -228,12 +226,6 @@ public class BalChangeConsumer implements DebeziumEngine.ChangeConsumer<ChangeEv
         if (returnValue instanceof BError) {
             ((BError) returnValue).printStackTrace();
         }
-    }
-
-    private BError createMethodNotFoundError(Payload payload, String methodName) {
-        BMap<BString, Object> detail = getEventProcessingErrorDetail(payload.toString());
-        return createError(EVENT_PROCESSING_ERROR, "Function '" + methodName + "' is not available.",
-                null, ValueCreator.createRecordValue(getModule(), EVENT_PROCESSING_ERROR_DETAIL, detail));
     }
 
     private boolean isHeartbeatEvent(ChangeEvent<String, String> record) {

--- a/native/src/main/java/io/ballerina/lib/cdc/Listener.java
+++ b/native/src/main/java/io/ballerina/lib/cdc/Listener.java
@@ -11,7 +11,7 @@
  * Unless required by applicable law or agreed to in writing,
  * software distributed under the License is distributed on an
  * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
- * KIND, either express q or implied. See the License for the
+ * KIND, either express or implied. See the License for the
  * specific language governing permissions and limitations
  * under the License.
  */
@@ -36,6 +36,9 @@ import io.debezium.engine.format.Json;
 
 import java.io.IOException;
 import java.math.BigDecimal;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
 import java.time.Instant;
 import java.time.temporal.ChronoUnit;
 import java.util.Map;
@@ -227,6 +230,8 @@ public class Listener {
 
     private static Object startEngine(Environment environment, BObject listener, Properties engineProperties,
                                       Long livenessInterval) throws Exception {
+        createParentDirectoriesIfNeeded(engineProperties);
+
         @SuppressWarnings("unchecked")
         ConcurrentHashMap<String, Service> serviceMap = (ConcurrentHashMap<String, Service>) listener
                 .getNativeData(TABLE_TO_SERVICE_MAP_KEY);
@@ -365,6 +370,22 @@ public class Listener {
             return createCdcError("Failed to invoke the liveliness check for the Debezium engine: " + e.getMessage());
         } finally {
             lock.unlock();
+        }
+    }
+
+    private static void createParentDirectoriesIfNeeded(Properties engineProperties) throws IOException {
+        String[] fileProps = {
+                "offset.storage.file.filename",
+                "schema.history.internal.file.filename"
+        };
+        for (String prop : fileProps) {
+            String filePath = engineProperties.getProperty(prop);
+            if (filePath != null) {
+                Path parent = Paths.get(filePath).getParent();
+                if (parent != null) {
+                    Files.createDirectories(parent);
+                }
+            }
         }
     }
 

